sdk/core/Azure.Core/src/DynamicData/DynamicData.AllowList.cs

```diff
@@ -0,0 +1,231 @@
+            {
+                if (value == null)
+                {
+                    return true;
+                }
+
+                if (!type.IsGenericType)
+                {
+                    return false;
+                }
+
+                if (type.GetGenericTypeDefinition() != typeof(Dictionary<,>))
+                {
+                    return false;
+                }
+
+                Type[] types = type.GetGenericArguments();
+                if (types[0] != typeof(string))
+                {
+                    return false;
+                }
+
+                if (IsAllowedPrimitive(types[1]))
+                {
+                    return true;
+                }
+
+                return IsAllowedEnumerableValue(types[1], ((IDictionary)value).Values);
+            }
+
+            private static bool IsAllowedEnumerableValue(Type elementType, IEnumerable enumerable)
+            {
+                foreach (var item in enumerable)
+                {
+                    if (item == null)
+                    {
+                        continue;
+                    }
+
+                    // Don't allow heterogenous collections
+                    if (item.GetType() != elementType)
+                    {
+                        return false;
+                    }
+
+                    if (!IsAllowedValue(item))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private static bool IsAllowedAnonymousValue<T>(Type type, T value, HashSet<Type> ancestorTypes)
+            {
+                if (!IsAnonymousType(type))
+                {
+                    return false;
+                }
+
+                foreach (PropertyInfo property in type.GetProperties())
+                {
+                    if (IsAllowedLeafType(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    object? propertyValue = property.GetValue(value);
+                    if (IsAllowedCollectionValue(property.PropertyType, propertyValue))
+                    {
+                        continue;
+                    }
+
+                    // Detect cycles: trust but verify
+                    if (ancestorTypes.Contains(property.PropertyType))
+                    {
+                        continue;
+                    }
+
+                    // Recurse
+                    ancestorTypes.Add(type);
+                    if (!IsAllowedAnonymousValue(property.PropertyType, propertyValue, ancestorTypes))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            private static bool IsAnonymousType(Type type)
+            {
+                return type.Namespace == null && type.Name.StartsWith("<>f__AnonymousType");
```

Comment:
I would remove the namespace check.

Reply:
If you use ordinal string comparison, StartsWith will return false very quickly for types that don't start with '<', which will be almost all types but anonymous types. i.e. do type.Name.StartsWith(..., StringComparison.Ordinal). See https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L978